Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,604 @@
+            (cast_pos.mpr hn0)
+        rw [← rpow_natCast, ← rpow_add hL0', ← neg_div, div_add_div_same, Nat.cast_add,
+          add_neg_cancel_right, Nat.cast_mul, ← rpow_mul (apply_nonneg f _), mul_one_div,
+          mul_div_assoc, rpow_mul (le_of_lt hL0')]
+        exact rpow_lt_rpow (apply_nonneg f _) hm1 h_lt
+      have h2 : f (x ^ (n % m1)) ^ (1 / (n : ℝ)) ≤ (f x ^ (n % m1)) ^ (1 / (n : ℝ)) := by
+        by_cases hnm1 : n % m1 = 0
+        · simpa [hnm1, pow_zero] using rpow_le_rpow (apply_nonneg f _) hf1 (one_div_cast_nonneg _)
+        · exact rpow_le_rpow (apply_nonneg f _) (map_pow_le_pow f _ hnm1) (one_div_cast_nonneg _)
+      have h3 : (L + ε / 2) * (L + ε / 2) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) *
+            (f x ^ (n % m1)) ^ (1 / (n : ℝ)) ≤ L + ε := by
+        have heq : L + ε = L + ε / 2 + ε / 2 := by rw [add_assoc, add_halves]
+        have hL0' : 0 < L + ε / 2 := add_pos_of_nonneg_of_pos hL0 (half_pos hε)
+        rw [heq, ← tsub_le_iff_left]
+        nth_rw 3 [← mul_one (L + ε / 2)]
+        rw [mul_assoc, ← mul_sub, mul_comm, ← le_div_iff₀ hL0', div_div]
+        exact hm2 n (le_trans (le_max_right (m1 : ℕ) m2) hn)
+      have h4 : 0 < f (x ^ (n % ↑m1)) ^ (1 / (n : ℝ)) := rpow_pos_of_pos hxn' _
+      have h5 : 0 < (L + ε / 2) * (L + ε / 2) ^ (-(↑(n % ↑m1) / (n : ℝ))) :=
+        mul_pos hL0' (rpow_pos_of_pos hL0' _)
+      calc f (x ^ ((m1 : ℕ) * (n / (m1 : ℕ)) + n % m1)) ^ (1 / (n : ℝ)) =
+            f (x ^ ((m1 : ℕ) * (n / (m1 : ℕ))) * x ^ (n % m1)) ^ (1 / (n : ℝ)) := by rw [pow_add]
+        _ ≤ (f (x ^ ((m1 : ℕ) * (n / (m1 : ℕ)))) * f (x ^ (n % m1))) ^ (1 / (n : ℝ)) :=
+          (rpow_le_rpow (apply_nonneg f _) (map_mul_le_mul f _ _) (one_div_cast_nonneg _))
+        _ = f (x ^ ((m1 : ℕ) * (n / (m1 : ℕ)))) ^ (1 / (n : ℝ)) *
+            f (x ^ (n % m1)) ^ (1 / (n : ℝ)) :=
+          (mul_rpow (apply_nonneg f _) (apply_nonneg f _))
+        _ ≤ (f (x ^ (m1 : ℕ)) ^ (n / (m1 : ℕ))) ^ (1 / (n : ℝ)) *
+              f (x ^ (n % m1)) ^ (1 / (n : ℝ)) := ((mul_le_mul_right h4).mpr h)
+        _ < (L + ε / 2) * (L + ε / 2) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) *
+              f (x ^ (n % m1)) ^ (1 / (n : ℝ)) := (mul_lt_mul h1 (le_refl _) h4 (le_of_lt h5))
+        _ ≤ (L + ε / 2) * (L + ε / 2) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) *
+              (f x ^ (n % m1)) ^ (1 / (n : ℝ)) := ((mul_le_mul_left h5).mpr h2)
+        _ ≤ L + ε := h3
+
+/-- If `f 1 ≤ 1`, then `smoothingSeminorm' f x` is the limit of `smoothingSeminormSeq f x`
+  as `n` tends to infinity. -/
+theorem tendsto_smoothingSeminorm' (hf1 : f 1 ≤ 1) (x : R) :
+    Tendsto (smoothingSeminormSeq f x) atTop (𝓝 (smoothingSeminorm' f x)) := by
+  by_cases hx : f x = 0
+  · exact tendsto_smoothingSeminorm'_of_eq_zero f hx
+  · exact tendsto_smoothingSeminorm'_of_ne_zero f hf1 hx
+
+/-- If `f 1 ≤ 1`, then `smoothingSeminorm' f x` is nonnegative. -/
+theorem smoothingSeminorm_nonneg (hf1 : f 1 ≤ 1) (x : R) : 0 ≤ smoothingSeminorm' f x := by
```

Comment:
A prime in `smoothingSeminorm` is lacking (which brings me back at the question about its existence).